### ISSUE TITLE
Improve detection of dead connections in the client

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#4641]
This is a follow up of https://github.com/rethinkdb/rethinkdb/issues/4572 .

The default settings for TCP keepalive on Linux mean that we won't detect connection loss until 11 minutes have passed.

> http://stackoverflow.com/a/14855726/2413884 has an overview of how to set the available options from Python on Linux and OS X, and it also has a link to Windows' documentation on the topic.
> 
> It looks like we could tune the timeout parameters on Python, though it wouldn't be portable (i.e. we would need different code for different platforms).

> In the long term, we can either implement our own protocol-level heartbeat, or add code to set the TCP keepalive parameters on a per-platform basis. For node.js there might be a way of doing it with https://www.npmjs.com/package/raw-socket .

**Comment by danielcompton** [comment]
What is the solution for 3rd party drivers? Do we also need to enable keepalive?

**Comment by danielmewes** [comment]
Yes, I believe in Go you can use setsockopt similar to the Python change.

**Comment by danielmewes** [comment]
@danielcompton Ahm not sure why I mentioned Go to you. I must have confused you with @dancannon when I wrote that email. I'm not sure how it works in Clojure.
